<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>BIDistances</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">BIDistances</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BIDistances)</span></code></pre></div>
<div id="introduction-to-bioinformatic-distances" class="section level1">
<h1>Introduction to Bioinformatic Distances</h1>
<p>This packages contains various functions for distances-measures
useful for bioinformatic data.</p>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>Installation using GitHub</p>
<p>#<code>{r} #library(remotes) #install_github(&quot;Mthrun/BIDistances&quot;) #</code></p>
</div>
<div id="examples" class="section level1">
<h1>Examples</h1>
<div id="cosinusdistance" class="section level2">
<h2>CosinusDistance</h2>
<p>The cosine distance is a distance-measure based on the cosine
similarity. Let <span class="math inline">\(A\)</span> be the data
matrix and <span class="math inline">\(A_i\)</span>, <span class="math inline">\(A_j\)</span> some row vectors of <span class="math inline">\(A\)</span>. The cosine similarity is then defined
as <span class="math inline">\(\begin{equation} \text{s(i,j)} =
\cos(\theta) = \frac{\mathbf{A_i} \cdot \mathbf{A_j}}{|\mathbf{A_i}|
|\mathbf{A_j}|} \end{equation}\)</span>, and the cosine distance as
<span class="math inline">\(d(i,j)=\max{s}-s(i,j)\)</span>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Hepta) </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>distMatrix <span class="ot">=</span> <span class="fu">CosinusDistance</span>(Hepta<span class="sc">$</span>Data)</span></code></pre></div>
</div>
<div id="dist2all" class="section level2">
<h2>Dist2All</h2>
<p>The Dist2All function calculates the distances of a given point <span class="math inline">\(x\)</span>, to all other points (rows) of a given
data matrix <span class="math inline">\(A\)</span>. For the calculation
of the distances, various distance-measures can be chosen, for
e.g. Euclidean, Manhattan (City Block), Mahalanobis, Bhjattacharyya, for
a complete list see <a href="https://CRAN.R-project.org/package=parallelDist">parallelDist</a>.
The distance-measure can be specified with the method argument. The
function returns an ordered vector of the distances from point <span class="math inline">\(x\)</span> to all points in <span class="math inline">\(A\)</span> in ascending order, as well as the
indices of k-nearest-neighbors for the chosen distance measure.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Hepta)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>V <span class="ot">=</span> <span class="fu">Dist2All</span>(Hepta<span class="sc">$</span>Data[<span class="dv">1</span>,],Hepta<span class="sc">$</span>Data, <span class="at">method =</span> <span class="st">&quot;euclidean&quot;</span>, <span class="at">knn=</span><span class="dv">3</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Vector of distances from Hepta$Data[1,] to all other rows in Hepta$Data</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(V<span class="sc">$</span>distToAll)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [38] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [75] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [112] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [149] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [186] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Vector of the indices of the k-nearest-neighbors, according to the euclidean distance</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(V<span class="sc">$</span>KNN)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 1 2 3</span></span></code></pre></div>
</div>
<div id="distancematrix" class="section level2">
<h2>DistanceMatrix</h2>
<p>For a given <span class="math inline">\([1:n, 1:d]\)</span> data
matrix <span class="math inline">\(A\)</span>, with <span class="math inline">\(n\)</span> cases and <span class="math inline">\(d\)</span> variables, the function calculates the
symmetric <span class="math inline">\([1:n, 1:n]\)</span> distance
matrix, given a chosen distance-measure. The method argument specifies
the distance-measure (euclidean by default).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Hepta)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>Dmatrix <span class="ot">=</span> <span class="fu">DistanceMatrix</span>(Hepta<span class="sc">$</span>Data, <span class="at">method=</span><span class="st">&#39;euclidean&#39;</span>)</span></code></pre></div>
<p>Options for method include :</p>
<p>‘euclidean’, ‘sqEuclidean’, ‘binary’, ‘cityblock’, ‘maximum’,
‘canberra’, ‘cosine’, ‘chebychev’, ‘jaccard’, ‘mahalanobis’, ‘minkowski’
,‘manhattan’ , ‘braycur’ ,‘cosine’.</p>
<p>For the method ‘minkowski’, the parameter dim, can be used to specify
the value of p in <span class="math inline">\(\left( \sum_{i=1}^{n}
|A_{j i} - A_{l i}|^p \right)^{1/p}\)</span></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>Dmatrix <span class="ot">=</span> <span class="fu">DistanceMatrix</span>(Hepta<span class="sc">$</span>Data, <span class="at">method=</span><span class="st">&#39;minkowski&#39;</span>, <span class="at">dim=</span><span class="dv">3</span>)</span></code></pre></div>
</div>
<div id="fractional-distances" class="section level2">
<h2>Fractional Distances</h2>
<p>The fractional distance function uses the formula of the
Minkowski-metric to calculate the distances and allows the usage of
fractional values <span class="math inline">\(p \in [0,1]\)</span>,
which can be useful for high-dimensional data [Aggrawal et al.,
2001].</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Hepta)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>distMatrix <span class="ot">=</span> <span class="fu">FractionalDistance</span>(Hepta<span class="sc">$</span>Data, <span class="at">p =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>)</span></code></pre></div>
</div>
<div id="tfidf-distance" class="section level2">
<h2>Tfidf-distance</h2>
<p>The term frequency-inverse document frequency (Tf-idf) is a
statistical measure of relevance of a term <span class="math inline">\(t\)</span> to a document <span class="math inline">\(d\)</span> in a collection of documents <span class="math inline">\(D\)</span>. The Tfidf-distance for two documents
<span class="math inline">\(d_i\)</span>, <span class="math inline">\(d_j \in D\)</span> is then the absolute difference
between the Tfidf-values.</p>
<p>An exemplary usage for bioinformatic data is the calculation of
distances between genes using the Tfidf-distance, based on GO-Terms
(Gene-Ontology-terms). For this a matrix <span class="math inline">\(A\)</span> of <span class="math inline">\(n\)</span> genes as rows and <span class="math inline">\(m\)</span> GO-Terms as columns is used, where
genes can be interpreted as documents and GO-terms as terms [Thrun,
2022].</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Hearingloss_N109)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>V <span class="ot">=</span> <span class="fu">Tfidf_dist</span>(Hearingloss_N109<span class="sc">$</span>FeatureMatrix_Gene2Term, <span class="at">tf_fun =</span> mean)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get distances</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>dist <span class="ot">=</span> V<span class="sc">$</span>dist</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Get weights</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>TfidfWeights <span class="ot">=</span> V<span class="sc">$</span>TfidfWeights</span></code></pre></div>
<p>For the calculation of the (augmented) term-frequency, per default
the mean of the non-zero entries is used, but can be specified with the
argument tf_fun.</p>
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>[Thrun, 2021] Thrun, M. C.: The Exploitation of Distance
Distributions for Clustering, International Journal of Computational
Intelligence and Applications, Vol. 20(3), pp. 2150016, DOI:
10.1142/S1469026821500164, 2021.</p>
<p>[Thrun, 2022] Thrun, M. C.: Knowledge-based Indentification of
Homogenous Structures in Genes, 10th World Conference on Information
Systems and Technologies (WorldCist’22), in: Rocha, A., Adeli, H.,
Dzemyda, G., Moreira, F. (eds) Information Systems and Technologies,
Lecture Notes in Networks and Systems, Vol 468.,pp. 81-90, DOI:
10.1007/978-3-031-04826-5_9, Budva, Montenegro, 12-14 April, 2022.</p>
<p>[Aggrawal et al., 2001] Aggrawal, C. C., Hinneburg, A., Keim, D.
(2001), On the Suprising Behavior of Distance Metrics in High
Dimensional Space.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
